aip/general/0236.md

```diff
@@ -0,0 +1,355 @@
+- If multiple previews are supported, calling the create more than once on a resource **should** lead to the creation of multiple experiments represented as new subresources until the quota is reached.
+- If the operation failed, return the appropriate error code ([AIP-193][aip-193]).
+
+[aip-133]: https://aip.dev/133
+[aip-193]: https://aip.dev/193
+
+#### update
+
+```proto
+rpc UpdatePolicyExperiment(UpdatePolicyExperimentRequest) returns (PolicyExperiment);
+```
+- Standard methods: Update ([AIP-134]) **must** be used to change the fields-under-test in an experiment used for preview mode.
+  - The name inside “policy” **must not** be changed.
+- The Control Plane **should** set the PreviewMetadata.state to INACTIVE if the state was ACTIVE at the time of an update. This is so the user can easily distinguish between previews of different fields-under-test.
+- If the operation failed, the old experiment **must** be returned alongside the appropriate error code ([AIP-193][aip-193]).
+
+[aip-134]: https://aip.dev/134
+
+#### get
+
+```proto
+rpc GetPolicyExperiment(GetPolicyExperimentRequest) returns (PolicyExperiment);
+```
+- Standard methods: Get ([AIP-131][aip-131]) **must** be used to get the experiment used for preview mode.
+
+[aip-131]: https://aip.dev/131
+
+#### list
+
+```proto
+rpc ListPolicyExperiments(ListPolicyExperimentsRequest) returns (ListPolicyExperimentsResponse);
+```
+- Standard methods: List ([AIP-132][aip-132]) **should** be used to list experiments for preview mode.
+- List **should** support Filtering ([AIP-160][aip-160]) semantics.
+- Filtering on PreviewMetadata.state **should** indicate which experiments are actively being previewed or not.
+
+[aip-132]: https://aip.dev/132
+[aip-160]: https://aip.dev/160
+
+#### delete
+
+```proto
+rpc DeletePolicyExperiment(DeletePolicyExperimentRequest) returns (google.protobuf.Empty);
+```
+- Standard methods: Delete ([AIP-135][aip-135]) **must** be used to delete experiments.
+
+[aip-135]: https://aip.dev/135
+
+### Parent Resource API Changes
+
+#### delete
+
+- A delete of the top-level resource **must** delete all experiments.
+- To maintain the experiments, the top-level Policy **must** be changed to a no-op Policy instead of using delete.
+
+### Custom Method APIs
+
+#### startPreview
+
+```proto
+rpc StartPreviewPolicyExperiment(StartPreviewPolicyExperimentRequest) returns (StartPreviewPolicyExperimentResponse) {
+  option (google.api.http) = {
+    post: "/v1/{name=policies/*/policyExperiments/*}:startPreview"
+    body: "*"
+  }
+}
+
+message StartPreviewPolicyExperimentRequest {
+  string name = 1;
+  string etag = 2;
+}
+
+message StartPreviewPolicyExperimentResponse {
+  string name = 1;
+  string etag = 2;
+  string parent_etag = 3;
+  bool successful = 3;
+  string error_message = 4;
+}
+```
+- Admin flow API to start previewing a resource **should** be implemented.
+  - Remains declarative friendly as only an output only field is modified.
```

Comment:
Well, that's true the method is effectively invisible to desired-state systems. But that also means it's impossible to create an experiment that will produce information.

Consider adding a `bool start_automatically` field to the experiment if you would like to enable desired-state system users to be able to add an experiment that will produce logs by virtue of deploying.

```
data google_theservice_thepolicy "existingpolicy" {
  name = "my-policy-id"
}

resource google_theservice_thepolicy_experiment "trysomething" {
  parent = data.google_theservice_thepolicy.existingpolicy.id
  name = "my-experiment-id"
  start_automatically = true
  rules = [ ... ]
}
```